### ISSUE TITLE
Empty validation images

### ISSUE THREAD

**Issue by komninoschatzipapas** [bghira/SimpleTuner#360]
My validation images are coming out empty:
<img src="https://github.com/bghira/SimpleTuner/assets/29049104/d405aba9-b383-43bf-a2c6-bd5a2085ab36)" width="30%">
<img src="https://github.com/bghira/SimpleTuner/assets/29049104/cf557251-1581-42fb-b2e5-60188ce0503a)" width="30%">

Which is odd as I'm only doing one epoch on top of juggernaut. One of my suspicions is that I'm using the Juggernaut model but the VAE was defaulted to `madebyollin/sdxl-vae-fp16-fix`. I'm running another training with Juggernaut's VAE and will update the issue in ~ 1 hour. Any other help will be very much appreciated!

<details>
<summary>sdxl-env.sh</summary>
export SIMPLETUNER_LOG_LEVEL=DEBUG
export SIMPLETUNER_TRAINING_LOOP_LOG_LEVEL=DEBUG
export MODEL_TYPE='full'
export USE_BITFIT=false
export LEARNING_RATE=8e-7 #@param {type:"number"}
export MODEL_NAME="orionsoftware/juggernaut-XL-v9"

export DEBUG_EXTRA_ARGS="--report_to=wandb"
export TRACKER_PROJECT_NAME="sdxl-training"
export TRACKER_RUN_NAME="simpletuner-sdxl"

export NUM_EPOCHS=1
export MAX_NUM_STEPS=0

export DATALOADER_CONFIG="/workspace/multidatabackend.json"
export OUTPUT_DIR="/workspace/output"

export RESOLUTION=832
export RESOLUTION_TYPE="pixel"
export MINIMUM_RESOLUTION=$RESOLUTION

export VALIDATION_PROMPT="full body photo, woman, asian, 30 years old"
export VALIDATION_GUIDANCE=7.5
export VALIDATION_GUIDANCE_RESCALE=0.7
export VALIDATION_STEPS=100
export VALIDATION_NUM_INFERENCE_STEPS=30
export VALIDATION_SEED=42
export VALIDATION_RESOLUTION=$RESOLUTION

export TRAIN_BATCH_SIZE=4
export GRADIENT_ACCUMULATION_STEPS=1

export LR_SCHEDULE="polynomial"

export CAPTION_DROPOUT_PROBABILITY=0.03

export METADATA_UPDATE_INTERVAL=65
export VAE_BATCH_SIZE=$TRAIN_BATCH_SIZE

export DELETE_ERRORED_IMAGES=0
export DELETE_SMALL_IMAGES=0

export TRAINING_SCHEDULER_TIMESTEP_SPACING="trailing"
export INFERENCE_SCHEDULER_TIMESTEP_SPACING="trailing"

export MIN_SNR_GAMMA=5

export USE_XFORMERS=true

export USE_GRADIENT_CHECKPOINTING=true

export ALLOW_TF32=true
export OPTIMIZER="adamw_bf16"

export TRAINER_EXTRA_ARGS="${TRAINER_EXTRA_ARGS} --prediction_type=v_prediction --rescale_betas_zero_snr"
export TRAINER_EXTRA_ARGS="${TRAINER_EXTRA_ARGS} --training_scheduler_timestep_spacing=trailing --inference_scheduler_timestep_spacing=trailing"

export TRAINING_SEED=-1
export MIXED_PRECISION="bf16"
export PURE_BF16=true
export TRAINING_DYNAMO_BACKEND='no'
</details>

<details>
<summary>debug.log</summary>
<a>https://file.io/k67RAlR5zs3w</a>
</details>

**Comment by bghira** [comment]
what prediction type is it using? it will say this in the info logs

**Comment by bghira** [comment]
ah i see you did export the v-prediction type options but you can only do that on sd2.1 or terminus-xl models

**Comment by bghira** [comment]
to finish this training will probably take about 10k images and a few weeks is my guess.

**Comment by BuildBackBuehler** [comment]
I'm also having Validation Image issues. I get through 1 epoch no problem, 10 batchsize, 1 gradient accumulation step, getting ~100-120s per iteration. Epoch ends, pipeline loads 7/7, just freezing at Validation Images. No kill process, terminated the training myself after 10mins without any movement. 

I've run with a few different configs. --unet_attention_slice (produced NaaN on 2x pre-training), --enable_multiprocessing, no problems here, --vae_cache_preprocess (no problems here). Nothing too wild settings wise. 

Nothing we haven't discussed setting-wise that'd be out-of-the-ordinary. I did add --rescale_betas_zero_snr --lora_init_type "gaussian" --use_dora. Standard SDXL prediction-type...though maybe I should change that along w/ the SNR weight setting, but I need to read on optimal settings for those 2 heh.

OT, also, had to remove my 1200 max_training_steps (I had it as an arg. in combination with 5 epochs) because it was doing 1200 training steps per epoch for 1200 epochs. Right now the progress bar (at least) is displaying that there are only 5 steps per epoch. So that strikes me as a bug unless I'm missing something. I imagine it should have initially shown 5 epochs, 1200 training steps per epoch. Then the 2nd instance, the steps shouldn't have been dictated by the epochs, no?

**Comment by komninoschatzipapas** [comment]
@BuildBackBuehler - try setting the steps to 0 if you’re setting an epochs number. If you set the epochs without any number for the steps it does seem to bug out indeed.

**Comment by bghira** [comment]
DoRA is an experimental thing that isn't too well tested. unet attention slicing seems to be required in very few scenarios, so it's there, but it also seems problematic.

**Comment by komninoschatzipapas** [comment]
Removing v prediction fixed it for me. It would be nice if we could specify the height and width of each validation image.

Kohya allows you to append `--w 832 --h 1216 --l 5 --s 50` to your prompt and set the sampler via a cli option.

**Comment by bghira** [comment]
every feature added needs to be tested before release, so i don't do things there are other general solutions for, eg. using an external inference script from the toolkit dir.

**Comment by bghira** [comment]
![image](https://github.com/bghira/SimpleTuner/assets/59658056/bef310f0-36eb-41d6-a5a5-feafdd64df88)

it's available in the DeepFloyd branch `feature/deepfloyd` as an augmentation to `--validation_resolution` which now accepts the previous integer value as well as `WxH` or `WxH,int,int,WxH` comma-separated list of mixed simple and custom resolutions.

**Comment by BuildBackBuehler** [comment]
Hm so I unfortunately have not gotten to the bottom of this. 

Just removed --use_dora & --enable_multiprocessing with no luck. I was using it because of purported memory savings

Here's my logs if it helps.

https://api.wandb.ai/links/arthuraardvark/6eeqkuuw

@komninoschatzipapas I'll give the setting to 0 a try! I'd rather set the 2 to my desired #s (figure in an ideal world, the script would then opt for whichever was more sensible considering batch size/repeats/etc)

And OT but thank you BGhira for your dedication 💖 I like the aspect ratio rounding! I'll see if I can pull together the feature I'd suggested with a landscape/square/portrait buckets. Lest you think that'd cause more problems than problems solved which I figure may be why you introduced the more abstract/globally-apt solution of the rounding.

**Comment by bghira** [comment]
* don't rescale betas for zero snr unless you're *training a zero-snr model* like bytedance's stuff or my terminus series
* don't use a learning rate of 1.0 for AdamWBF16
* remove --use_prodigy_optimizer, it's not available right now
* would recommend avoiding DoRA until you have training any other way stabilised
* try default lora init method instead of gaussian
* would recommend 'leading' timestep spacing for training/inference on base SDXL and related models.

**Comment by komninoschatzipapas** [comment]
> ![image](https://private-user-images.githubusercontent.com/59658056/324204635-bef310f0-36eb-41d6-a5a5-feafdd64df88.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM2NTUxMTUsIm5iZiI6MTcxMzY1NDgxNSwicGF0aCI6Ii81OTY1ODA1Ni8zMjQyMDQ2MzUtYmVmMzEwZjAtMzZlYi00MWQ2LWE1YTUtZmVhZmRkNjRkZjg4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDIwVDIzMTMzNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZmNDIyMWEyNzBhMjY2N2UxNDVhNzVhY2I2NGZiOGEzZmVjZGU1ZWNlYjY2ZDk2NDYwMmQxYzY0YWFmZWNjY2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.iEVqCgkRUFz8Fj5G77VeP6w5-LaddpI7VN1Rou_Kfko)
> 
> it's available in the DeepFloyd branch `feature/deepfloyd` as an augmentation to `--validation_resolution` which now accepts the previous integer value as well as `WxH` or `WxH,int,int,WxH` comma-separated list of mixed simple and custom resolutions.

Legend!

**Comment by BuildBackBuehler** [comment]
> * don't rescale betas for zero snr unless you're _training a zero-snr model_ like bytedance's stuff or my terminus series
> * don't use a learning rate of 1.0 for AdamWBF16
> * remove --use_prodigy_optimizer, it's not available right now
> * would recommend avoiding DoRA until you have training any other way stabilised
> * try default lora init method instead of gaussian
> * would recommend 'leading' timestep spacing for training/inference on base SDXL and related models.

Agh, tried to hold off as long as I could, was so certain that the swap to opt AdamWBF16 would do the trick but I have gone through at least a dozen more runs without success. Actually took a step backwards because I'm not even able to get an epoch in. My #1 suspect right now is Diffusers, moreso damn Pipenv's crappy dependency logic. It has swapped out Torch/Torch Vision and others because the pre-release is not being recognized by packages that require Torch (and then doubles down with 2.2.2). Anyways, right now it *should* be using Beinzei's Sub-Attention branch of Diffusers and I *think* that's what got me through an epoch. But I'll swap it out. Accelerate may have gone from 29.2 to 29.3, Transformers from 4.39.x to 4.40.0.

Hanging at the start of the epoch, gives an OOM. Looking at the verbose Dynamo logs, it completes step 1 and then when it goes to TVM to (re?)compile for step 2 it was stating that TVM received some values it couldn't/shouldn't for Symbolic s0 thru s4 values (kinda hoping that Wandb saves the day and just saves alllll of my Torch Dynamo logs, too because I didn't heh). I had installed TVM as the backend myself, yesterday I had been using Eager, so I have been doing runs with Eager most recently, too. 

Other than Diffusers swap...maybe clearing all my caches or using PYTORCH_ENABLE_MPS_FALLBACK=1. Only other thought I have is that I believe I used an env. var. to suppress Dynamo errors and continue training when it was working (up to validation step). This time around, because I couldn't find what I input, I put the runtime's suggestion into play by inserting it, and some other Dynamo config ops I read about, at the beginning of the train_SDXL.py script...not sure if that's copacetic. 

These

'''
torch._dynamo.config.guard_nn_modules_using_dict_tags = False
torch._dynamo.config.repro_after = 'dynamo'
# torch._dynamo.config.optimize_ddp = 1
# torch._dynamo.config._ddp_optimization_mode = 'ddp_optimizer'
torch._dynamo.config.optimize_ddp_lazy_compile = True
torch._dynamo.config.raise_on_unsafe_aot_autograd = True
torch._dynamo.config.error_on_nested_fx_trace = False
torch._dynamo.config.suppress_errors = True
'''

And I was running with args. as per your instructions for the most part (no --use_dora, lesser extent no Gaussian noise). 60/40 without --enable_multiprocessing --vae_cache_preprocess, --validation_torch_compile=True.  

If wandb doesn't have the logs, I'll recreate 'em. 

https://wandb.ai/arthuraardvark/trainingfml/table?nw=nwuserarthuraardvark

**Comment by BuildBackBuehler** [comment]
Well, latest run finally jumped the Epoch 1 hurdle. I don’t know about the others, some I did wait a longggg time before canceling, but this 1 had Dynamo errors in Eager and 10 seconds afterward it got to generating! Of course only 5 steps on account of the 5 epochs 😅. 

I wish I hadn’t overlooked that proper VAE comment @komninoschatzipapas — that’s what did the trick I think. Also did have to forego DoRA.

But back to stuck on validation image generation. I’m going to let this generation and my Mac cook while I sleep. See if it ever trips up and produce some errors at least… 

A last ditch solution I figured (based on the success of random GitHub Kohya users who didn’t exactly strike me as knowledgeable lol) is to drop the Adam-bfloat16 to use Adam-8 bit. 

May just go back to my rig, try to use the Tesla M40 as the frontman, then offload everything to my RTX 3070 (+ Mac…if it works)

**Comment by bghira** [comment]
* sub-quad attention is only for inference, it doesn't help at training time.
* i'd use a venv just for simpletuner, and use `poetry install --no-root` to bring in the proper dependencies
* nothing but bf16 will (probably ever) work on Mac, adamwbf16 is really all you can use
* any other optimizer WILL fail because the weights aren't stored using fp32 precision levels
* disable all torch compile options

i really wish you'd just gone through the tutorial and used the basic setup instead of going knee-deep in DoRA and all the other stuff, there's no benefit to having things enabled when you don't know what they do or how they work. i will disable the torch compile things, that's my fault. it has seemingly never worked.
